### DISCUSSION TITLE
cs_tools will not install

### DISCUSSION THREAD

**Discussion by RandyHoffpauir** [thoughtspot/cs_tools#118] [General] [answered]
We have been using cs_tools as part of our azuredevop build process for several months now. It has been working great until today. It will no longer install using this:
(Invoke-WebRequest `
    -Uri https://raw.githubusercontent.com/thoughtspot/cs_tools/master/cs_tools/updater/_bootstrapper.py `
    -UseBasicParsing `
).Content | python - --install

It does not show any error and the log file for the install looks like below:
"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\2ff36240-c622-4c1b-9b89-cafef2b60b5a.ps1'"
21:26:01 | Welcome to the CS Tools Bootstrapper!
         | Ideally, you will only need to run this one time, and then your environment can be fully managed by CS Tools itself.
         | If you run into any issues, please reach out to us on GitHub Discussions below.
         | 
         |           GitHub: https://github.com/thoughtspot/cs_tools/issues/new/choose
         | 
21:26:02 | Missing '_updater.py', downloading from GitHub
         | Downloaded as '_updater.py'
         | Creating the CS Tools virtual environment.
Finishing: Install cs_tools

Clearly this is not a complete install but I don't have any error messages.
Any ideas?

**Comment by boonhapus** [comment]
Hey @RandyHoffpauir does this piece happen automatically? I don't recognize it as part of our install process.

```
"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a_temp\2ff36240-c622-4c1b-9b89-cafef2b60b5a.ps1'"
```

My thinking is that output might be getting redirected and that's why you don't see any useful logs.

**Reply by RandyHoffpauir** [reply]
Yes, it is happening automatically.  This is happening in azure devops which is usually pretty good at logging all the information. 

This is the beginning of what it looks like when it runs successfully:
"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\cfc644f6-e8a2-43d9-8a26-1ec1f3f6644b.ps1'"
14:50:51 | Welcome to the CS Tools Bootstrapper!
         | Ideally, you will only need to run this one time, and then your environment can be fully managed by CS Tools itself.
         | If you run into any issues, please reach out to us on GitHub Discussions below.
         | 
         |           GitHub: https://github.com/thoughtspot/cs_tools/issues/new/choose
         | 
         | Missing '_updater.py', downloading from GitHub
14:50:52 | Downloaded as '_updater.py'
         | Creating the CS Tools virtual environment.
14:51:03 | Requirement already satisfied: pip>=23.1 in c:\users\vssadministrator\appdata\roaming\cs_tools\.cs_tools\lib\site-packages (23.2.1)
14:51:04 | Collecting pip>=23.1
         | Obtaining dependency information for pip>=23.1 from https://files.pythonhosted.org/packages/8a/6a/19e9fe04fca059ccf770861c7d5721ab4c2aebc539889e97c7977528a53b/pip-24.0-py3-none-any.whl.metadata
         | Downloading pip-24.0-py3-none-any.whl.metadata (3.6 kB)
         | Downloading pip-24.0-py3-none-any.whl (2.1 MB)
         | ---------------------------------------- 2.1/2.1 MB 12.3 MB/s eta 0:00:00
         | Installing collected packages: pip
         | Attempting uninstall: pip
         | Found existing installation: pip 23.2.1
14:51:05 | Uninstalling pip-23.2.1:
         | Successfully uninstalled pip-23.2.1
14:51:09 | Successfully installed pip-24.0
         | Getting the latest CS Tools release.
14:51:10 | Found version: v1.4.16
         | Installing CS Tools and its dependencies.
14:51:11 | Collecting cs_tools@ https://github.com/thoughtspot/cs_tools/archive/v1.4.16.zip (from cs_tools[cli]@ https://github.com/thoughtspot/cs_tools/archive/v1.4.16.zip)
14:51:12 | Downloading https://github.com/thoughtspot/cs_tools/archive/v1.4.16.zip
14:51:14 | Installing build dependencies: started
14:51:20 | Installing build dependencies: finished with status 'done'
         | Getting requirements to build wheel: started
14:51:21 | Getting requirements to build wheel: finished with status 'done'
         | Installing backend dependencies: started

And it continues with a lot more install info from here

**Reply by boonhapus** [reply]
ack @RandyHoffpauir , I can reproduce this in a local environment now and am debugging now. Will update here once I fix or know more about what's going on.

**Reply by boonhapus** [reply]
Hey @RandyHoffpauir can you try the install again? I've added a change that should get around this issue for now!

**Reply by RandyHoffpauir** [reply]
It looks like that fixed it the install process but now I have another problem.
The next command to run is the following:
.\cs_tools.exe config create --config primary --username myusername --password mypassword --host https://mycompany.thoughtspot.cloud --disable_sso --verbose --default

But this is now getting the following error:
Id     Name            PSJobTypeName   State         HasMoreData     Location             Command                  
--     ----            -------------   -----         -----------     --------             -------                  
1      a7f4d581-056...                 NotStarted    False                                ...                      
[14:01:12] ERROR    Missing parameter: url                          main.py:111

**Reply by boonhapus** [reply, marked as answer]
Ah @RandyHoffpauir can you use `--url` instead of `--host`? I will add an alias for backwards compatibility on this one.

**Reply by RandyHoffpauir** [reply]
That got it thanks for the help!

**Reply by RandyHoffpauir** [reply]
I spoke too soon. Now when we try to do a commit by running this command:
.\cs_tools.exe tools git branches commit --tag mytag --branch-name mybranch --comment "mycomment" --delete-aware'
we are getting this error:
{
    'error': {
        'message': {
            'debug': {
                'code': 15025,
                'incident_id_guid': 'f3fb015a-8522-4943-aaa1-ff7b24b89386',
                'debug': 'Repository contains unmerged paths'
            }
        }
    }
}

**Reply by boonhapus** [reply]
Hmm I'm not 100% sure on this @RandyHoffpauir , but I believe this is a git/VCS API issue. Let me post this one internally and see what the resolution should be.

**Reply by boonhapus** [reply]
It might be helpful if you could provide the last few log files. 

`cs_tools logs report --latest 5`

**Reply by boonhapus** [reply]
Hey @RandyHoffpauir based on some internal JIRAs, it's looking like this is an issue with the VCS APIs. Do you mind raising a ThoughtSpot support ticket with error message and getting help there? I took a stab at ghostwriting, feel free to copy/paste if you find it useful.

> Hi,
>
> We are calling the `api/rest/2.0/vcs/git/branches/commit` API and receiving the following error.
>
> I understand what the debug is saying, but have no indication on how I would go about resolving this. How can I fix this?
> 
> ```
> {
>     'error': {
>         'message': {
>             'debug': {
>                 'code': 15025,
>                 'incident_id_guid': 'f3fb015a-8522-4943-aaa1-ff7b24b89386',
>                 'debug': 'Repository contains unmerged paths'
>             }
>         }
>     }
> }
> ```

**Reply by RandyHoffpauir** [reply]
Thanks for the help. I was able to fix it by calling api/rest/2.0/vcs/git/config/delete to remove my config, then calling api/rest/2.0/vcs/git/config/create to create it again. No idea why this works but it does.

**Reply by boonhapus** [reply]
😅 This is the resolution I found internally as well @RandyHoffpauir , however was unsure if that's the recommended approach. I see on some related JIRAs here at ThoughtSpot that this case is planned to be handled natively by TS version 9.12.0 , so hopefully you won't have to run this kind of "fix" for much longer.

**Comment by tdasmoh** [comment]
I am also having the same problem. I am on a Mac OS. When I installed I didn't get any error messages however when I run a cs_tools command, I keep getting command not found: cs_tools.
Could you please help me with this?
![Uploading Screenshot 2024-04-04 at 2.18.26 PM.png…]()

**Reply by boonhapus** [reply]
I think your image didn't load @tdasmoh . Have you tried restarting the terminal once you make it through install?

We add a bit to your shell profile in order to enable the `cs_tools` command globally, and this only takes effect once you restart the terminal window.

**Reply by boonhapus** [reply]
Hey @tdasmoh can you try the install again? I've added a change that should get around this issue for now!

**Reply by tdasmoh** [reply]
Hi @boonhapus I tried both restarting terminal and installing again just now. Still getting the same issue :(

**Reply by boonhapus** [reply]
@tdasmoh would you be free to meet on Zoom? I'll hop on to take a look with you. Please shoot me an email and we can schedule a time to get you past this one.
